pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -311,23 +377,42 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 
 	scopedLog = scopedLog.WithField(logfields.EndpointID, ep.StringID())
 
-	targetServerAddr, err := p.lookupTargetDNSServer(w)
+	targetServerIP, targetServerPort, targetServerAddr, err := p.lookupTargetDNSServer(w)
 	if err != nil {
-		scopedLog.WithError(err).Error("Cannot extract target server address to forward DNS request to")
-		stat.Err = fmt.Errorf("Cannot extract target server address to forward DNS request to: %s", err)
+		log.WithError(err).Error("cannot extract destination IP:port from DNS request")
+		stat.Err = fmt.Errorf("Cannot extract destination IP:port from DNS request: %s", err)
 		stat.ProcessingTime.End(false)
 		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, stat)
 		p.sendRefused(scopedLog, w, request)
 		return
 	}
-	scopedLog.WithField("server", targetServerAddr).Debug("Found target server to of DNS request")
+
+	serverSecID, exists, err := p.LookupSecIDByIP(targetServerIP)
+	if !exists || err != nil {
+		scopedLog.WithError(err).WithField("server", targetServerAddr).Debug("cannot find server ip in ipcache")
+		stat.Err = fmt.Errorf("Cannot find server ip in ipcache: %s", err)
+		stat.ProcessingTime.End(false)
+		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, stat)
+		p.sendRefused(scopedLog, w, request)
```

Comment:
The ordering of the above few statements for this error handling case is different from the equivalent cases just a few lines down. Is there any particular reason? (Mostly just concerned about consistency of timestamps and so on)

Reply:
good catch! Processing errors are reported one way whereas blocks are treated another (namely, the sendRefused processing time and error are included in NotifyOnDNSMsg). The .End(true/false) also needs to be consistent.
I've fixed things up in this file (I hope)